File: graph-java/src/main/java/com/cc/graph/gep/Chromosome.java
package com.cc.graph.gep;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.concurrent.ThreadLocalRandom;
import java.util.stream.Collectors;

import com.cc.graph.base.Graph;

public class Chromosome {

    public final Set<Gene> genes;

    public Chromosome(Set<Gene> genes) {
        this.genes = Collections.unmodifiableSet(genes.stream()
                .filter(e -> e.size() > 0).collect(Collectors.toSet()));
    }

    public static Chromosome generate(Graph graph) {
        List<String> nodeIds = new ArrayList<>(graph.vertexIds());
        ThreadLocalRandom random = ThreadLocalRandom.current();
        Collections.shuffle(nodeIds, random);
        Set<Gene> temps = new HashSet<Gene>();
        while (nodeIds.size() > 0) {
            int size = nodeIds.size();
            int position;
            if (size > 5) {
                position = random.nextInt(3, size);
                if (size - position <= 2) {
                    position = size;
                }
                temps.add(new Gene(new HashSet<String>(nodeIds.subList(0,
                        position))));
                nodeIds = nodeIds.subList(position, size);
            } else {
                temps.add(new Gene(new HashSet<String>(nodeIds)));
                nodeIds = Collections.emptyList();
            }
        }
        return new Chromosome(temps);
    }

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + ((genes == null) ? 0 : genes.hashCode());
        return result;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj)
            return true;
        if (obj == null)
            return false;
        if (getClass() != obj.getClass())
            return false;
        Chromosome other = (Chromosome) obj;
        if (genes == null) {
            if (other.genes != null)
                return false;
        } else if (!genes.equals(other.genes))
            return false;
        return true;
    }

    public List<Set<String>> toCommunityStyle() {
        return genes.stream().map(g -> g.nodes).collect(Collectors.toList());
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        for (Gene g : genes) {
            sb.append(g.toString());
        }
        return sb.toString();
    }
}

File: graph/src/main/scala/com/cc/graph/gep/Chromosome.scala
package com.cc.graph.gep

import com.cc.graph.base.Graph
import java.util.concurrent.{ ThreadLocalRandom => TLRandom }
import util.Random

case class Chromosome private (genes: Set[Gene]) {
  override def toString(): String = genes.mkString

  def toCommunityStyle: Graph.Community = genes.map(_.nodes).toList
}

object Chromosome {

  /**
   * generate a chromosome of the graph nodes<br/>
   * randomly choose a position( >= 2) and make the nodes before the position a community
   */
  def generate(graph: Graph): Chromosome = {
    val nodeIds = Random.shuffle(graph.vertexIds.toVector)
    def gen(ids: Vector[String], genes: Set[Gene]): Chromosome = {
      val idSize = ids.size
      if (idSize == 0) {
        Chromosome(genes)
      } else {
        val position = if (idSize > 3) TLRandom.current().nextInt(3, ids.size) else 3
        val (take, rest) = ids.splitAt(position)
        if (rest.size >= 3) {
          gen(rest, genes + Gene(take: _*))
        } else {
          Chromosome(genes + Gene(ids: _*))
        }
      }
    }
    gen(nodeIds, Set())
  }

  /**
   * construct chromosome <br/>
   * the gene whose node size is zero will be removed
   */
  def apply(gs: Gene*): Chromosome = {
    Chromosome(gs.filterNot(_.nodes.size == 0).toSet)
  }
}


